### DISCUSSION TITLE
rmarkdown syntax for tabset

### DISCUSSION THREAD

**Discussion by Pierre9344** [quarto-dev/quarto-cli#2449] [Q&A] [answered]
Hello,

Is it possible to use the rmardown syntax "## tab title {.tabset}" to generate tabset? It would be useful to generate tabset procedurally.
I tried the code bellow found on [stackoverflow](https://stackoverflow.com/questions/57314417/rmarkdown-define-header-with-tabset-option-dynamically-inside-loop) but it didn't generate tabset.

````
```{r, echo = FALSE, results='asis'}
headers <- c("membership", "participation")
tabs <- c("age", "gender")

for (i in headers){

    cat("###", i, "{.tabset}", "\n")
    cat("\n\n\n")

    for (j in tabs) {

        cat("####", j, "\n")
        cat("\n\n\n") 
    }
} 
```
````

I can use the script bellow to generate tabset for the first level (h3 title) but the next levels (h4) are not converted to tabset

````
::: {.panel-tabset}

```{r}
#| results='asis'
for (i in 1:nrow(ls)) {
  cat("### ", ls$VAR[i], "\n")
  load(paste0("./data/base/proposalClassification/4visits/", permut, " permut/", ls$files[i]))
  #plot(cluster_model)
  cluster_assignments <- get_clusters(cluster_model, number_clusters = 2) %>%
    dplyr::mutate(assignment_num_clus_2 = ifelse(.$assignment_num_clus_2 == 1, "A", "B")) %>%
    dplyr::rename(!!ls$VAR[i] := assignment_num_clus_2)
  colnames(df)[1] <- "ID" # rename the column v1_id
  if(i == 1) {
    original_data <- df %>% left_join(cluster_assignments, by = "ID")
  } else {
    original_data <- original_data %>% left_join(cluster_assignments, by = "ID")
  }


  cat("\n")
}
```
:::
````

**Comment by mcanouil** [comment, marked as answer]
You have to use an extra backticks to format your document, _i.e._, four in your case ` ```` `.

Short answer to the question: no, you can't.

Tabset feature in Quarto is not implemented exactly as in `rmarkdown`.
Every level 2 headings in a cell div `::: {.panel-tabset}` will be considered as a tabset, which is compatible with any procedural to write code/text as tabset (probably using `knitr::knit_child()` and/or `#| output: asis`).
Depending on your format, please read the documentation and look at the example:
- https://quarto.org/docs/output-formats/html-basics.html#tabsets
- https://quarto.org/docs/presentations/revealjs/index.html#tabsets

**Reply by Pierre9344** [reply]
Thank you for the answer. I'm not sure on how to use `knitr::knit_child` but I have found a solution by using the cat function to print the `::: {.panel-tabset}` and `:::` parts

Here is an example if someone need it:

````
```{r}
#| output: asis
for (i in c("A","B","C")) {
  cat(paste0("### ", i, "\n"))
  cat("::: {.panel-tabset} \n")
  for (j in c("X","Y")) {
    cat(paste0("#### ", j, "\n"))
    plot(1)
    cat("\n")
  }
  cat("\n:::\n")
}
```
````

**Reply by mcanouil** [reply]
`knitr::knit_child()` is useful if you generate a code cell in a code cell and wat it to be computed.
In your case, it's "plain" text/output.

See for example: https://github.com/quarto-dev/quarto-cli/discussions/2094#discussioncomment-3465685

**Reply by cderv** [reply]
See also this specific thread with examples of how to programmatically create tabs and nested tabs : https://github.com/quarto-dev/quarto-cli/issues/2370

`knit_child` is to be used when you need to add R code content to include (like `plot(1)` above). It is safer to do it with `knit_child()` that mixing with `cat()` IMO. We documented that in the R Markdown Cookbook, and the recipies apply to Quarto as well. 

See also https://hbiostat.org/rflow/rformat.html#quarto-report-writing-helper-functions for the `maketabs` function that someone has made for this kind of usage I believe.
